core/lib/tee_verifier/src/lib.rs

```diff
@@ -7,60 +7,28 @@
 use std::{cell::RefCell, rc::Rc};
 
 use anyhow::Context;
-use serde::{Deserialize, Serialize};
 use zksync_crypto::hasher::blake2::Blake2Hasher;
 use zksync_merkle_tree::{
-    BlockOutputWithProofs, TreeInstruction, TreeLogEntry, TreeLogEntryWithProof,
+    BlockOutputWithProofs, TreeInstruction, TreeLogEntry, TreeLogEntryWithProof, ValueHash,
 };
 use zksync_multivm::{
-    interface::{FinishedL1Batch, L1BatchEnv, L2BlockEnv, SystemEnv, VmInterface},
+    interface::{FinishedL1Batch, L2BlockEnv, VmInterface},
     vm_latest::HistoryEnabled,
     VmInstance,
 };
-use zksync_object_store::{serialize_using_bincode, Bucket, StoredObject};
-use zksync_prover_interface::inputs::{PrepareBasicCircuitsJob, StorageLogMetadata};
+use zksync_prover_interface::inputs::{
+    PrepareBasicCircuitsJob, StorageLogMetadata, TeeVerifierInput, V1TeeVerifierInput,
+};
 use zksync_state::{InMemoryStorage, StorageView, WriteStorage};
 use zksync_types::{block::L2BlockExecutionData, L1BatchNumber, StorageLog, H256};
 use zksync_utils::bytecode::hash_bytecode;
 use zksync_vm_utils::execute_tx;
 
-/// Version 1 of the data used as input for the TEE verifier.
-#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
-pub struct V1TeeVerifierInput {
-    prepare_basic_circuits_job: PrepareBasicCircuitsJob,
-    l2_blocks_execution_data: Vec<L2BlockExecutionData>,
-    l1_batch_env: L1BatchEnv,
-    system_env: SystemEnv,
-    used_contracts: Vec<(H256, Vec<u8>)>,
-}
-
-/// Data used as input for the TEE verifier.
-#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
-#[non_exhaustive]
-#[allow(clippy::large_enum_variant)]
-pub enum TeeVerifierInput {
-    /// `V0` suppresses warning about irrefutable `let...else` pattern
-    V0,
-    V1(V1TeeVerifierInput),
+pub trait Verifiable {
+    fn verify(self) -> anyhow::Result<(ValueHash, L1BatchNumber)>;
 }
 
-impl TeeVerifierInput {
-    pub fn new(
-        prepare_basic_circuits_job: PrepareBasicCircuitsJob,
-        l2_blocks_execution_data: Vec<L2BlockExecutionData>,
-        l1_batch_env: L1BatchEnv,
-        system_env: SystemEnv,
-        used_contracts: Vec<(H256, Vec<u8>)>,
-    ) -> Self {
-        TeeVerifierInput::V1(V1TeeVerifierInput {
-            prepare_basic_circuits_job,
-            l2_blocks_execution_data,
-            l1_batch_env,
-            system_env,
-            used_contracts,
-        })
-    }
-
+impl Verifiable for TeeVerifierInput {
```

Comment:
`TeeVerifierInput` is an API envelope that has entries that cannot be verified. Moreover, implementing the trait for the envelope introduces double indirection (interface + enum fields) which doesn't make much sense.
It's better to implement this trait for `V1TeeVerifierInput` and handle `TeeVerifierInput::V0` separately in the prover itself.